### PR TITLE
05 fit

### PR DESCRIPTION
**Pull request by Yoshanuikabundi**
I've spent a lot of time investigating performance and getting things working on ROCm. My findings are that most of the time of any given batch/step is copying the batch dataset to the GPU. So I think we could get some substantial improvement in both throughput and efficiency by using multiple GPUs - each GPU could hold on to its portion of the dataset between epochs, data could be written to the GPU once at the start of the fit, and then each epoch would only consist of compute instead of compute and copying 30 gigs of unchanging dataset back and forth between CPU and GPU. 

Failing that, it might be possible to compress the dataset into batches once at the start of the training, send each compressed batch to the GPU one at a time, and then decompress and process it on the GPU.

I also found that PyTorch would use caching to use all available VRAM, even though each batch fits in relatively small amount of memory. This cache can be cleared at the end of each batch (though this is not implemented), but it cuts performance in half (making my GPU slower than my CPU), so you're better off just setting a VRAM limit if you need to use the GPU for, like, graphics at the same time as doing a fit.

### Help please!

I've taken advantage of the small size of the tetramers dataset to do a full, thousand epoch fit locally. Took about 3 hours. I'm fairly convinced I've done something wrong because:

- The loss doesn't improve much
- The energy portion of the loss remains at 0.0 for the entire run
- The proper torsions in the fitted force field have an awful lot of 0.0s

Would be great to get more eyes on this!

image

### CODE REVIEW

**Comment by lilyminium** [comment]
I've only just now taken a quick look, it's a bit hard without the dataset you're training to, but upshot is I think something slipped by in the dataset creation phase. Some dot points below:
* the proper torsions having a lot of 0s isn't very alarming with this training -- the tetramers is a dataset with very limited diversity that I'd only expect to touch a fairly limited set of torsions. As long as all torsions used in the tetramers have vaguely non-zero force constants, it looks like that part's going right and everything that should be trained is being trained (you almost definitely already know this, but just in case, you can profile which torsions are in a molecule with `ff.label_molecules(mol.to_topology())[0]["ProperTorsions"]`)
* Not too sure what to think of the loss, other than that yes I'd expect a faster and higher degree of improvement followed by plateauing, but this is likely to do with the below
* The energy loss being 0 is extremely worrisome. I think that's because you're training only to the set of QCA data (and it's lucky you did or this error would likely not have been noticed -- nice work!) The [QCA data treats each conformer of a molecule singly](https://github.com/openforcefield/back-to-school-josh/blob/94c64fa26dcc34aa37a07ba005f8a4608033b4a4/workflow/02-select-data/__main__.py#L240-L276) but a descent Entry actually requires [all conformers to be grouped together](https://simonboothroyd.github.io/descent/latest/reference/targets/energy/#descent.targets.energy.Entry). I believe this has been done with SPICE (especially since there's ~25k systems) but could be worth checking that also. My bad, I really should have caught this, especially since there were logs written out -- instead of expecting 1k data points, we should expect 200 data points (molecules), each with 5 conformers. Grouping the conformers together is important for learning relative energies. Sorry!

**Comment by Yoshanuikabundi** [comment]
Second fit with grouped conformers: 
image

Forces are going the wrong way...

**Comment by lilyminium** [comment]
I'm confused that the total loss doesn't seem to be the sum of the force and energy loss... edit: no it is, I'm just bad at maths

You could weight their contributions to the total loss to force the optimizer to work with the forces more as well?

**Comment by j-wags** [comment]
FWIW I wondered whether the energy vs. force discrepancy was due to a unit conversion issue when formatting the QC data, so I combed through your [code to do that](https://github.com/openforcefield/back-to-school-josh/blob/94c64fa26dcc34aa37a07ba005f8a4608033b4a4/workflow/02-select-data/__main__.py#L178-L193) for 30ish minutes and didn't find anything problematic (though this is assuming that energies and forces in SPICE are in hartree and hartree/bohr respectively, which does appear to be the case).
